### ISSUE TITLE
Hit tab and slide vertically or horizontally if required

### ISSUE THREAD

**Issue by juandahveed** [alvarotrigo/fullPage.js#4027]
### Description
basically, I hooked into the onLeave() callback and disabled the tab key and then hooked into the afterLoad() callback and enabled the tab key. I wrote custom accessibility to help users use a tab key / screen reader, etc.

my code does fire however, when I preventDefault() on the tab key, it does nothing. The goal is to disable tabs while the sections scroll, and then enable them again.

If you hit tab while the sections scroll it breaks fullpage.js.

However it works perfectly if you slowly hit tab and wait until sections finish loading.

Thanks in advance.

Here's a site to see it on.

http://staging.ddc-law.flywheelsites.com/

also, here's the code I've written so you can see it in full

** update ** --- It works perfectly for slide, but not section

`(function ($, window, undefined) {

    var scrollingTabBlocker = function (event) {
        // should stop tab during transition
        if (event.keyCode === 9 || event.key === 'Tab') {
             event.preventDefault();
         }
        
        var active = $('#fullpage > .active');
        var completely = $('#fullpage > .fp-completely');
       
        if (event.keyCode === 9 || event.key === 'Tab') {
        
             if (active.html() === completely.html()) {
        //         console.log('tab pressed and transitioning');
                 event.preventDefault();
             }
       
        }

    }

    function initFullpage() {

        new fullpage('#fullpage', {
            //options here
            licenseKey: 'F710F678-C1D84836-A6FBFDFA-7AABA375',

            // navigation
            menu: '#fullPageMenu',
            navigationTooltips: ['Top', 'Our Focus', 'Personal Injury', 'Transactional Law', 'About Our Firm',
                'Proven Results', 'Practice Areas', 'Location'],
            showActiveTooltip: true,
            navigation: true,
            slidesNavigation: true,

            // scrolling
            scrollingSpeed: 1250,
            autoScrolling: true,
            scrollHorizontally: false,
            loopHorizontal: true,

            // design
            responsiveWidth: 769,
            responsiveHeight: 750,
            paddingTop: '84px',
            verticalCentered: true,

            // Accessibility
            keyboardScrolling: true,
            animateAnchor: true,
            recordHistory: true,

            //events
            afterRender: function () {
                //on page load, start the slideshow
                slideTimeout = setInterval(function () {
                    fullpage_api.moveSlideRight();
                }, 4500);
            },

            onSlideLeave: function (section, origin, destination, direction) {

                if ($(window).width() < 768 && $slider) {

                    if (section.index == 0 && origin.index == 0 && direction == 'right' || direction == 'left') {
                        return false;
                    }
                }

                // cancel tab events during "transition" ** slides
                window.addEventListener('keydown', scrollingTabBlocker, true);
                window.addEventListener('keyup', scrollingTabBlocker, true);

            },

            afterSlideLoad: function (anchorLink, index, slideAnchor, slideIndex) {
                // allow tab events after "transition" ** slides
                window.removeEventListener('keydown', scrollingTabBlocker, true)
                window.removeEventListener('keyup', scrollingTabBlocker, true)
            },

            onLeave: function (index, nextIndex, direction) {
                // cancel tab events during "transition" ** sections
                console.log('stop tab events while moving');
                window.addEventListener('keydown', scrollingTabBlocker, false);
                window.addEventListener('keyup', scrollingTabBlocker, false);

            },

            afterLoad: function (anchorLink, index) {
                // allow tab events after "transition" ** sections
                console.log('allow tab events after moving');
                window.removeEventListener('keydown', scrollingTabBlocker, false)
                window.removeEventListener('keyup', scrollingTabBlocker, false)

            }

        });

        $slides = $('.section:not(.slide1)');
        $firstSlide = $('.section > .wp-block-group__inner-container');

        $($slides).each(
            function () {
                $(this).children('.wp-block-group__inner-container').addClass('top-level-slide-inner-container');
                $(this).children().children('.wp-block-group__inner-container').addClass('top-level-slide-inner-container');

                if ($(window).width() <= 768) {
                    $(this).children().addClass('fp-auto-height');
                    $(this).children().children().addClass('fp-auto-height');
                    $('.section.slide1').children().children('.fp-slides').addClass('fp-auto-height');
                }

            }
        );

        $('.slides').find('> .wp-block-group__inner-container').addClass('fp-auto-height');
        $('.slide5 > .fp-tableCell').addClass('fp-auto-height');

        // move last slide content into footer, then move footer into last slide
        var $lastSlide = $('.slide8');
        var $footer = $('body.home footer.page-footer').addClass('top-level-slide-inner-container'); // added class
        var $lastSLideBlockContent = $lastSlide.children().children('.wp-block-group__inner-container');

        $lastSLideBlockContent.detach().prependTo($footer);
        $footer.detach().appendTo($lastSlide.children());

    }

    function initSlick() {
        $slider = $('.practice-area-grid');
        $initialized = $('.practice-area-grid.slick-initialized').length > 0 ? true : false;

        if ($(window).width() <= 768 && $slider) {
            // make a slider
            $slider.slick({
                appendArrows: '.practice-area-grid',
                prevArrow: '<button type="button" class="slick-prev">PREV</button>',
                nextArrow: '<button type="button" class="slick-next">NEXT</button>'
            });
        } else {
            // destroy a slider
            if ($slider && $initialized) {
                $slider.slick('unslick');
            }

        }

        var logo_carousel = $('.logo-carousel');
        if (logo_carousel.length) {
            logo_carousel.slick({
                dots: true,
                slidesToShow: 5,
                slidesToScroll: 5,
                arrows: false,
                responsive: [
                    {
                        breakpoint: 1300,
                        settings: {
                            slidesToShow: 4,
                            slidesToScroll: 4,
                            infinite: true,
                            dots: true
                        }
                    },
                    {
                        breakpoint: 1024,
                        settings: {
                            slidesToShow: 3,
                            slidesToScroll: 3
                        }
                    },
                    {
                        breakpoint: 768,
                        settings: {
                            slidesToShow: 2,
                            slidesToScroll: 2
                        }
                    },
                    {
                        breakpoint: 480,
                        settings: {
                            slidesToShow: 1,
                            slidesToScroll: 1
                        }
                    }
                ]
            });
        }
    }

    initFullpage();
    initSlick();

    // lastly let's have some type of resize function in place, because... javascript
    window.onresize = function () {
        initFullpage();
        initSlick();
    }

    // make sure all of fullPage.js has focusable elements
    $('#fullpage').find('a, button, input').each(function (index, el) {
        // lets make sure everything has a tabindex
        $(el).attr('tabindex', "0");
    });

    // we want the side nav not focusable
    $('#fp-nav').find('a').each(function (index, el) {
        // lets make sure everything has a tabindex
        $(el).attr('tabindex', "-1");
    });

    var trueCount = 0;

    var accessibilityTabKeyStrokeHelpers = function (event) {
        var activeElement = document.activeElement;
        var activeSection = fullpage_api.getActiveSection();

        var active = $('#fullpage > .active');
        var completely = $('#fullpage > .fp-completely');

        // check if tab key
        if (event.keyCode === 9 || event.key === 'Tab') {

            // safety net to stop tab during transition
            if (active.get(0) !== completely.get(0)) {
                event.preventDefault();
            }

            if (activeElement.id === 'fullpage') {
                // this is the full page element, move to the first slide button
                document.getElementById('hero_button_1').focus();
            }

            if (activeElement.children[0] && activeElement.children[0].innerText === 'Proven Results') {
                // last item on first section, move to section 2
                fullpage_api.moveTo(2);
            }

            if (activeSection.index !== 0 && event.defaultPrevented === true) {
                trueCount = trueCount + 1;
                // if this isn't slide 1 and isn't the last slide, when the event is defaultPrevented = true we move
                // to the next slide
                if (trueCount > 1) {
                    trueCount = 0;
                    fullpage_api.moveSectionDown();
                }
            }

            if (activeSection.isLast === true) {
                // this is the last slide

                if (activeElement.ariaLabel === "Accessibility Statement") {
                    // make sure focus is "Accessibility Statement"
                    // go back to start
                    // reset document.focus() back to the very beginning
                    fullpage_api.moveTo(1);
                    activeElement.blur();
                    document.body.focus();
                }

            }
        }
    }

    window.addEventListener('keydown', accessibilityTabKeyStrokeHelpers, false)

})(jQuery, this);
`

**Comment by alvarotrigo** [comment]
That's a big code.
But basically, you are saying you found a bug in fullPage.js that makes it break when hitting tab while the transition takes place?

Could you please provide an isolated reproduction in [jsfiddle](https://jsfiddle.net/alvarotrigo/ea17skjr/) or [codepen](https://codepen.io/alvarotrigo/pen/qqabrp), with no CSS or JS files external to fullPage.js and the minimum amount of HTML code? Use empty sections unless strictly necessary for the reproduction.

If I'm able to reproduce it I'll mark it as a bug so we can investigate it for the next release.

**Comment by juandahveed** [comment]
I love that, I think I just need to be able to remove the named event listener.... So if the event listeners were variables, and I could reach them through the fullpage_api, I could manually remove the event listener on one callback and then manually add the event listener again with another callback. I'll try to set all this up for you later tonight.

**Comment by juandahveed** [comment]
@alvarotrigo  here is a link to the fiddle, I did see that it seems to work fine in firefox but not in Chrome. Let me know what you 
think and why that could be the case?

If you tab slowly through the sections, the elements will show their focus and when you hit the last element of each section it will transition to the next section. When you hit the last element of the last section it will go back to the first section. As long as you don't tab while sections are transitioning it works fine, but if you "jam on the tab" key while it is moving sections, you will see the issue i'm trying to resolve. You'll also see that my event listeners are being added / removed and that explains why I'm having difficulties. Let me know if I can help anymore. I did also have 2 external files but they were only the fullPage.js css and js files. I tried to keep everything as small as possible.

https://codepen.io/juandahveed/pen/bGpVyoO

**Comment by alvarotrigo** [comment]
Thanks for that!
I think we are talking about different things here. 

I was asking for an isolated reproduction where I can see if there's in fact a bug in fullpage.js.
However, it seems that what you want is to find a solution to your problem where you want fullPage.js to slide up or down by hitting tab if there are focusable elements instead of keeping its default behaviour, which is not moving to any other section/slide.

So here's an isolated reproduction without your extra code where we can see how fullPage.js prevents tab on any element outside the current section/slide. And it seems to work as expected.
https://codepen.io/alvarotrigo/pen/eYZJYJg?editors=1010

I'll change the title of the issue and mark it as a possible enhancement. 
We can perhaps add an option in fullpage.js to decide if we want to only tab on the current slide or to scroll vertically and horizontally when required. 

Regarding your current problem, I guess you can try asking in other places as this is more of a generic JavaScript question than a fullPage.js specific one.

**Comment by alvarotrigo** [comment]
> Regarding your current problem, I guess you can try asking in other places as this is more of a generic JavaScript question than a fullPage.js specific one.

What about not removing the tab event? Not sure why you remove them :)

**Comment by juandahveed** [comment]
> Thanks for that!
> I think we are talking about different things here.
> 
> I was asking for an isolated reproduction where I can see if there's in fact a bug in fullpage.js.
> However, it seems that what you want is to find a solution to your problem where you want fullPage.js to slide up or down by hitting tab if there are focusable elements instead of keeping its default behaviour, which is not moving to any other section/slide.
> 
> So here's an isolated reproduction without your extra code where we can see how fullPage.js prevents tab on any element outside the current section/slide. And it seems to work as expected.
> https://codepen.io/alvarotrigo/pen/eYZJYJg?editors=1010
> 
> I'll change the title of the issue and mark it as a possible enhancement.
> We can perhaps add an option in fullpage.js to decide if we want to only tab on the current slide or to scroll vertically and horizontally when required.
> 
> Regarding your current problem, I guess you can try asking in other places as this is more of a generic JavaScript question than a fullPage.js specific one.

**Comment by juandahveed** [comment]
> > Regarding your current problem, I guess you can try asking in other places as this is more of a generic JavaScript question than a fullPage.js specific one.
> 
> What about not removing the tab event? Not sure why you remove them :)

This is a great question, so essentially, I have it checking to see if it's the last focus-able on the section, then to go to the next section, OR if it's the last section and the last focus-able, then start over.... But you understand all that.

What I was trying to accomplish by removing the tab event, is that if you remove the tab event on the onLeave() call back, and then add the tab event back on the afterLoad() callback, then in-between those two callback you have disabled the tab event. So in theory, if the sections are moving and someone hits the tab key a bunch of times, nothing will break and the sections will continue to move to their proper places. Then when the section is fully loaded, the tab event will come back, allowing you to tab through all the focus-able items.

**Comment by alvarotrigo** [comment]
> and then add the tab event back on the afterLoad() callback, then in-between those two callback you have disabled the tab event.

I guess the problem is there yeah. I've tested it again on the [default fullpage.js behaviour](https://codepen.io/alvarotrigo/pen/eYZJYJg?editors=1010) and it seems it also breaks sometimes if i scroll and then hit tab at the same time. 
But I guess this case is a much less probably issue than in your particular scenario.

**Comment by alvarotrigo** [comment]
> So in theory, if the sections are moving and someone hits the tab key a bunch of times, nothing will break and the sections will continue to move to their proper places.

The problem here is that it seems the default tab behaviour is not prevented. It seems fullPage.js isn't preventing it on certain cases while scrolling, thus causing the issue.

**Comment by juandahveed** [comment]
> > So in theory, if the sections are moving and someone hits the tab key a bunch of times, nothing will break and the sections will continue to move to their proper places.
> 
> The problem here is that it seems the default tab behaviour is not prevented. It seems fullPage.js isn't preventing it on certain cases while scrolling, thus causing the issue.

Bingo, that's exactly it!!!!
This is where I was going when I was talking about extending the fullpage_api. Possibly having those event listeners use a named variable and not an anonymous function call if they are. And if that's the case, then I could manually remove the event listeners that I'm trying to override ( if that's what's happening here ) and then I can add the event listener again after the section is in place.

So, something like...

in the onLeave() callback --> fullpage_api.removeEventListener('defaultTab;);

and then......

in the afterLoad() callback --> fullpage_api.addEventListener('defaultTab;);

Does that make sense? Something like that.

**Comment by alvarotrigo** [comment]
But the problem is not on the event listeners used by fullPage.js. Those just prevent the default tab behaviour. 

It seems to me the problem is that fullPage.js isn't able to prevent the default tab behaviour used by the browser under those circumstances and therefore it tries to tab to the next focusable element while the page is being scrolled causing the issue.

You can even try to remove the whole `onTab` function in fullpage.js code and test again. But i think that won't solve it.

**Comment by juandahveed** [comment]
> But the problem is not on the event listeners used by fullPage.js. Those just prevent the default tab behaviour.
> 
> It seems to me the problem is that fullPage.js isn't able to prevent the default tab behaviour used by the browser under those circumstances and therefore it tries to tab to the next focusable element while the page is being scrolled causing the issue.
> 
> You can even try to remove the whole `onTab` function in fullpage.js code and test again. But i think that won't solve it.

Dang @alvarotrigo  I think you're right. That would explain why it works in Firefox and not Chrome. I'm guessing it's exactly the same in edge since it's the same engine. What a sticky situation. I'm not really sure where to go from here.

**Comment by dancingmonkey** [comment]
> Thanks for that! I think we are talking about different things here.
> 
> I was asking for an isolated reproduction where I can see if there's in fact a bug in fullpage.js. However, it seems that what you want is to find a solution to your problem where you want fullPage.js to slide up or down by hitting tab if there are focusable elements instead of keeping its default behaviour, which is not moving to any other section/slide.
> 
> So here's an isolated reproduction without your extra code where we can see how fullPage.js prevents tab on any element outside the current section/slide. And it seems to work as expected. https://codepen.io/alvarotrigo/pen/eYZJYJg?editors=1010
> 
> I'll change the title of the issue and mark it as a possible enhancement. We can perhaps add an option in fullpage.js to decide if we want to only tab on the current slide or to scroll vertically and horizontally when required.
> 
> Regarding your current problem, I guess you can try asking in other places as this is more of a generic JavaScript question than a fullPage.js specific one.

Setting tabindex="0" on every anchor tag totally fixed the page from not breaking for me. Also when I didn't set it on `<div id="fullpage" tabindex="0">` It didn't work so for anyone still experiencing this problem don't forget that one including all the anchor tags.
